src/System.Threading.Tasks/tests/System.Runtime.CompilerServices/AsyncTaskMethodBuilderTests.cs

```diff
@@ -527,6 +527,44 @@ public static void CancellationDoesntResultInEventFiring()
             TaskScheduler.UnobservedTaskException -= handler;
         }
 
+        [Fact]
+        public static async Task AsyncMethodsDropsStateMachineAndExecutionContextUponCompletion()
+        {
+            // Create a finalizable object that'll be referenced by both an async method's
+            // captured ExecutionContext and its state machine, invoke the method, wait for it,
+            // and then hold on to the resulting task while forcing GCs and finalizers.
+            // We want to make sure that holding on to the resulting Task doesn't keep
+            // that finalizable object alive.
+
+            var tcs = new TaskCompletionSource<bool>(TaskCreationOptions.RunContinuationsAsynchronously);
+
+            Task t = null;
+            await Task.Run(delegate // avoid any issues with the stack keeping the object alive
+            {
+                async Task YieldOnceAsync(object s)
+                {
+                    await Task.Yield();
+                    GC.KeepAlive(s); // keep s referenced by the state machine
+                }
+
+                var state = new InvokeActionOnFinalization { Action = () => tcs.SetResult(true) };
+                var al = new AsyncLocal<object> { Value = state }; // ensure the object is stored in ExecutionContext
+                t = YieldOnceAsync(state); // ensure the object is stored in the state machine
+                al.Value = null;
+            });
+
+            await t; // wait for the async method to complete and clear out its state
+
+            for (int i = 0; i < 2; i++)
+            {
+                GC.Collect();
+                GC.WaitForPendingFinalizers();
+            }
+
+            await tcs.Task; // finalizable object should have been collected and finalized
```

Comment:
>await tcs.Task; // fi [](start = 12, length = 21)

what happen if we have a bug and the object is not finalized? does this will hang? should we time out and fail the test instead?

Reply:
It would hang.  I can add a timeout.